flang/lib/Optimizer/Dialect/FIROps.cpp

```diff
@@ -2109,10 +2121,13 @@ static mlir::LogicalResult verify(fir::ReboxOp op) {
   if (inputEleTy != outEleTy)
     // TODO: check that outBoxTy is a parent type of inputBoxTy for derived
     // types.
-    // Character input and output types may be different if there is a
-    // substring in the slice, otherwise, they must match.
-    if (!inputEleTy.isa<fir::RecordType>() &&
-        !(op.getSlice() && inputEleTy.isa<fir::CharacterType>()))
+    // Character input and output types with constant length may be different if
+    // there is a substring in the slice, otherwise, they must match. If any of
+    // the types is a character with dynamic length, the other type can be any
+    // character type.
+    if (!(inputEleTy.isa<fir::RecordType>() ||
```

Comment:
It might be a little clearer to get rid of the dangling `!` condition/negative logic and swap the `return` statements.

Reply:
Yes, it is not the most readable. Switching the logic bothers me because someone adding additional checks might miss that a previous code path may also lead to an `return mlir::success()` before the added check, so I prefer keeping a single `return mlir::success();` at the end. Instead, I have named the "positive" condition `typeCanMismatch` and used `if (!typeCanMismatch) return op.emitOpError`.